feat(transformer): add `/* @_PURE__ */` annotation for pure calls

Dunqing opened the pull request:
close: #6072 

Many tests outputs are updated because we need to print `/* @_PURE__ */` for the pure calls.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9424?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9424** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9424?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9424?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9417** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9417?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9424).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
It seems like we need to turn off comments in conformance codegen.

Dunqing commented:
> It seems like we need to turn off comments in conformance codegen.

You mean turn off pure annotation comments? We need this to test if we added pure correctly. I found a lot of tests failed due to the way it is printed is not the same. For example:

```diff
 var HelloMessage = React.createClass({ render: function() {
   return /* @__PURE__ */ React.createElement("div", null, "Hello ", this.props.name);
 } });
-React.render(
-  /* @__PURE__ */ React.createElement(HelloMessage, { name: /* @__PURE__ */ React.createElement("span", null, "Sebastian") }),
-  mountNode
-);
+React.render(/* @__PURE__ */ React.createElement(HelloMessage, { name: /* @__PURE__ */ React.createElement("span", null, "Sebastian") }), mountNode);
```

I will dig into this problem later.

Dunqing commented on tasks/transform_conformance/snapshots/babel.snap.md:
```diff
@@ -2626,10 +2625,27 @@ after transform: ["T", "f"]
+x Output mismatch
+
+
+# babel-plugin-transform-react-jsx (133/145)
 * autoImport/after-polyfills-compiled-to-cjs/input.mjs
 x Output mismatch
 
+* pure/false-default-pragma-automatic-runtime/input.js
+x Output mismatch
+
+* pure/false-default-pragma-classic-runtime/input.js
+x Output mismatch
```
I have looked into Babel codebase, but I still haven't figured out why these two tests failed.

Dunqing commented on tasks/transform_conformance/snapshots/babel.snap.md:
```diff
@@ -2626,10 +2625,27 @@ after transform: ["T", "f"]
 
 
-# babel-plugin-transform-react-jsx (135/145)
+# babel-preset-react (6/9)
+* preset-options/pure/input.js
+x Output mismatch
+
+* preset-options/runtime-classic-pragma-no-frag/input.js
+x Output mismatch
+
+* regression/11294/input.mjs
+x Output mismatch
```
The first test fails as we don't have support `@babel/plugin-transform-react-pure-annotations`. The last two tests fail as `preset-react`'s `pure` option only defaults to enable `@babel/plugin-transform-react-pure-annotations`, not include enabling `@babel/transform-react-jsx`'s pure option. Currently both preset and plugin options are defined together, we don't have a way to only turn on `@babel/plugin-transform-react-pure-annotations` but turn off `@babel/transform-react-jsx`'s pure option

Dunqing replied:
See https://github.com/babel/babel/blob/578ac4df1c8a05f01350553950dbfbbeaac013c2/packages/babel-preset-react/src/index.ts#L36-L61

Dunqing commented on tasks/transform_conformance/src/test_case.rs:
```diff
@@ -311,17 +314,23 @@ impl TestCase {
                     CodeGenerator::new()
                         .with_options(CodegenOptions {
                             comments: false,
+                            // Disable pure annotation comments for async_to_generator plugin,
+                            // because it's weird some tests have it and some don't.
+                            annotation_comments: !babel_options.plugins.async_to_generator,
```
AsyncToGenerator is weird to print `pure`, I don't know why, so I disable checking `AsyncToGenerator` plugin for now

Dunqing commented on tasks/transform_conformance/src/test_case.rs:
```diff
@@ -311,17 +314,23 @@ impl TestCase {
                         })
                         .parse();
 
+                    // Clear comments to avoid pure annotation comments that cause mismatch.
+                    ret.program.comments.clear();
```
Fixed https://github.com/oxc-project/oxc/pull/9424#issuecomment-2689731729

In the transformer, we only set pure field to true, but the output has the real pure comment, which causes the output mismatch.

Boshen reviewed (approved):
Feel free to merge after fixing the conformance issues.

Dunqing commented:
> Feel free to merge after fixing the conformance issues.

Fixed in #9530

Dunqing committed:
feat(transformer): add `/* @_PURE__ */` annotation for pure calls (#9424)

close: #6072

Many tests outputs are updated because we need to print `/* @_PURE__ */` for the pure calls.
